### PR TITLE
PLAT-390: deduplication using VFindCallRequest

### PR DESCRIPTION
**Pull request by aalekseev-ru**
**- What I did**
add deduplication via VFindCallRequest
**- How I did it**
add new steps, rewrite old steps
**- How to verify it**
run unit tests, wait for plat-389 and run integration tests

### CODE REVIEW

File: ledger-core/virtual/execute/execute.go [hunk 1]
```diff
@@ -388,15 +391,97 @@ func (s *SMExecute) stepDeduplicateUsingPendingsTable(ctx smachine.ExecutionCont
...
-		// TODO: do we have result? if we have result then we should resend result
-		ctx.Log().Warn("duplicate found as pending request")
-		return ctx.Stop()
+		if isActive {
+			ctx.Log().Warn("duplicate found as pending request")
+			return ctx.Stop()
+		}
 	}
 
 	if ctx.AcquireForThisStep(pendingListFilled).IsNotPassed() {
```
Comment (sexyhamster) [inline ledger-core/virtual/execute/execute.go:400]: in your PR you also need to drop pendingListFilled semaphores

File: ledger-core/virtual/execute/execute.go [hunk 2]
```diff
@@ -388,15 +391,97 @@ func (s *SMExecute) stepDeduplicateUsingPendingsTable(ctx smachine.ExecutionCont
...
+	s.messageSender.PrepareAsync(ctx, func(goCtx context.Context, svc messagesender.Service) smachine.AsyncResultFunc {
+		err := svc.SendRole(goCtx, &msg, node.DynamicRoleVirtualExecutor, s.execution.Object, s.pulseSlot.PulseData().PrevPulseNumber())
+		return func(ctx smachine.AsyncResultContext) {
+			if err != nil {
+				ctx.Log().Error("failed to send message", err)
+			}
+		}
+	}).WithoutAutoWakeUp().Start()
+
+	return ctx.Sleep().ThenJump(s.stepProcessFindCallResponse)
```
Comment (sexyhamster) [inline ledger-core/virtual/execute/execute.go:451]: probably, you need to add stepWaitFindCallResponse, you could be woken up by different signal. And you need to for s.VFindCallResponse != nil.

File: ledger-core/virtual/execute/execute.go [hunk 3]
```diff
@@ -388,15 +391,97 @@ func (s *SMExecute) stepDeduplicateUsingPendingsTable(ctx smachine.ExecutionCont
...
+	case payload.FoundCall:
+		if s.VFindCallResponse.CallResult == nil {
+			ctx.Log().Warn("request found on previous executor, but there was no result")
+			return ctx.Stop()
+		}
+
+		ctx.Log().Warn("request found on previous executor, resend result")
+
+		target := s.Meta.Sender
+		s.messageSender.PrepareAsync(ctx, func(goCtx context.Context, svc messagesender.Service) smachine.AsyncResultFunc {
```
Comment (sexyhamster) [inline ledger-core/virtual/execute/execute.go:465]: maybe split stepSendCallResult in prepare and send, so here you can jump to send?

File: ledger-core/virtual/execute/execute.go [hunk 4]
```diff
@@ -388,15 +391,97 @@ func (s *SMExecute) stepDeduplicateUsingPendingsTable(ctx smachine.ExecutionCont
 	}
 
 	if isDuplicate {
-		// TODO: do we have result? if we have result then we should resend result
-		ctx.Log().Warn("duplicate found as pending request")
-		return ctx.Stop()
+		if isActive {
```
Comment (sexyhamster) [inline ledger-core/virtual/execute/execute.go:394]: if isDuplicate && isActive ?

File: ledger-core/virtual/execute/execute.go [hunk 5]
```diff
@@ -372,7 +375,7 @@ func (s *SMExecute) stepDeduplicateUsingPendingsTable(ctx smachine.ExecutionCont
 
 		if pendingList.Exist(s.execution.Outgoing) {
```
Comment (sexyhamster) [inline ledger-core/virtual/execute/execute.go:376]: maybe switch to 
```
isActive, exists := list.GetState(ref)
```

File: ledger-core/virtual/execute/execute.go (hunk 4 above)
Comment (aalekseev-ru) [reply ledger-core/virtual/execute/execute.go:387]: with SMAwaitTableFill ?

File: ledger-core/virtual/execute/execute.go (hunk 1 above)
Comment (aalekseev-ru) [reply ledger-core/virtual/execute/execute.go]: with SMAwaitTableFill ?

Comment (sexyhamster) [reply ledger-core/virtual/execute/execute.go]: yep. In deduplication with VE(P-1) there is no need to wait for full pending table

File: ledger-core/virtual/execute/execute.go (hunk 3 above)
Comment (aalekseev-ru) [reply ledger-core/virtual/execute/execute.go:446]: I thought about it, but it was a mess and I decided to do it this way

File: ledger-core/virtual/execute/execute.go (hunk 2 above)
Comment (aalekseev-ru) [reply ledger-core/virtual/execute/execute.go]: yeah, miss it

File: ledger-core/virtual/execute/execute.go [hunk 6]
```diff
@@ -387,14 +378,96 @@ func (s *SMExecute) stepDeduplicateUsingPendingsTable(ctx smachine.ExecutionCont
...
+				}
+			}
+		}).WithoutAutoWakeUp().Start()
+
+		return ctx.Jump(s.stepFinishRequest)
+
+	case payload.MissingCall:
+	case payload.UnknownCall:
+		// execute it
+		break
```
Comment (sexyhamster) [inline ledger-core/virtual/execute/execute.go:467]: go has break by default. This break is not required

**Comment by aalekseev-ru** [comment]
forgot to regen uml -_-

File: ledger-core/virtual/execute/execute.go [hunk 7]
```diff
@@ -70,6 +70,8 @@ type SMExecute struct {
 
 	delegationTokenSpec payload.CallDelegationToken
 	stepAfterTokenGet   smachine.SlotStep
+
+	VFindCallResponse *payload.VFindCallResponse
```
Comment (dgovorukhin) [inline ledger-core/virtual/execute/execute.go:74]: Why SM field is public?

File: ledger-core/virtual/execute/execute.go [hunk 8]
```diff
@@ -387,14 +380,101 @@ func (s *SMExecute) stepDeduplicateUsingPendingsTable(ctx smachine.ExecutionCont
...
+
+func (s *SMExecute) stepProcessFindCallResponse(ctx smachine.ExecutionContext) smachine.StateUpdate {
+	switch s.VFindCallResponse.Status {
+	case payload.FoundCall:
+		if s.VFindCallResponse.CallResult == nil {
+			ctx.Log().Warn("request found on previous executor, but there was no result")
+			return ctx.Stop()
+		}
+
+		ctx.Log().Warn("request found on previous executor, resend result")
```
Comment (dgovorukhin) [inline ledger-core/virtual/execute/execute.go:457]: Why warn log level?

File: ledger-core/virtual/execute/execute.go [hunk 9]
```diff
@@ -387,14 +380,101 @@ func (s *SMExecute) stepDeduplicateUsingPendingsTable(ctx smachine.ExecutionCont
...
+					ctx.Log().Error("failed to send message", err)
+				}
+			}
+		}).WithoutAutoWakeUp().Start()
+
+		return ctx.Jump(s.stepFinishRequest)
+
+	case payload.MissingCall:
+	case payload.UnknownCall:
+		// execute it
```
Comment (dgovorukhin) [inline ledger-core/virtual/execute/execute.go:473]: Log this on debug level

File: ledger-core/virtual/execute/execute.go (hunk 6 above)
Comment (ruz) [reply ledger-core/virtual/execute/execute.go]: done

File: ledger-core/virtual/execute/execute.go (hunk 7 above)
Comment (ruz) [reply ledger-core/virtual/execute/execute.go]: fixed

File: ledger-core/virtual/execute/execute.go (hunk 8 above)
Comment (ruz) [reply ledger-core/virtual/execute/execute.go]: switched to Trace

File: ledger-core/virtual/execute/execute.go (hunk 9 above)
Comment (ruz) [reply ledger-core/virtual/execute/execute.go:466]: done

File: ledger-core/virtual/execute/execute.go [hunk 10]
```diff
@@ -374,14 +367,105 @@ func (s *SMExecute) stepDeduplicateUsingPendingsTable(ctx smachine.ExecutionCont
...
+			}
+		}
+	}).WithoutAutoWakeUp().Start()
+
+	return ctx.Sleep().ThenJump(s.stepWaitFindCallResponse)
+}
+
+func (s *SMExecute) stepWaitFindCallResponse(ctx smachine.ExecutionContext) smachine.StateUpdate {
+	if s.findCallResponse == nil {
+		ctx.Sleep().ThenRepeat()
```
Comment (sexyhamster) [inline ledger-core/virtual/execute/execute.go:437]: Return

Comment (ruz) [reply ledger-core/virtual/execute/execute.go]: fixed

**Pull request merged**